### DISCUSSION TITLE
How to debug OAL script?

### DISCUSSION THREAD

**Discussion by zhyhzhylove** [apache/skywalking#8048] [Q&A]
I want to debug OAL, ex:
service_instance_relation_server_cpm = from(ServiceInstanceRelation.*).filter(detectPoint == DetectPoint.SERVER).cpm();

But not enter code in idea.

**Comment by wu-sheng** [comment]
What part do you expect to debug? OAL engine is a runtime code generator generally.

**Reply by zhyhzhylove** [reply]
I want to know how process data and when to save data, ex:
service_instance_relation_server_cpm = from(ServiceInstanceRelation.*).filter(detectPoint == DetectPoint.SERVER).cpm();

how cal ServiceInstanceRelation data and when save data into es.

When i know the whole process，i may know the specific meaning of the index and ext OAL engine to handle some special data for us.

OAL engine is an wonderful idea!!

**Reply by wu-sheng** [reply]
How to work is a code explanation which is out of community responsibility. 

If you want to know how it works, read from `MetricsStreamProcessor#create`. And system env `SW_OAL_ENGINE_DEBUG=Y` provided to output all generated binary classes into `oal-rt/` folder for developers.

**Reply by zhyhzhylove** [reply]
thanks.![image](https://user-images.githubusercontent.com/3204108/139803371-fcf1da5e-3831-4c72-bd13-849017e96823.png)
What is mean of functionName "cpm"?  
per min? I not find it in doc on v8.4, only read source code.

**Reply by wu-sheng** [reply]
CPM = Calls per minute

**Reply by zhyhzhylove** [reply]
thanks

**Reply by zhaozhihao59** [reply]
> How to work is a code explanation which is out of community responsibility.
> 
> If you want to know how it works, read from `MetricsStreamProcessor#create`. And system env `SW_OAL_ENGINE_DEBUG=Y` provided to output all generated binary classes into `oal-rt/` folder for developers.

I set SW_OAL_ENGINE_DEBUG=Y in my Idea but it doesn’t seem to take effect. There is no file in the target directory of oal-rt/. How should I deal with it? I set it when the program starts, and export SW_OAL_ENGINE_DEBUG=Y in my computer system environment

<img width="581" alt="image" src="https://github.com/apache/skywalking/assets/20415794/feb58a45-50b6-4079-a5b0-c9995c985b84">
